sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala

```diff
@@ -39,6 +42,23 @@ class MetricsReporter(
   registerGauge("processingRate-total", _.processedRowsPerSecond, 0.0)
   registerGauge("latency", _.durationMs.get("triggerExecution").longValue(), 0L)
 
+  private val timestampFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'") // ISO8601
+  timestampFormat.setTimeZone(DateTimeUtils.getTimeZone("UTC"))
+
+  registerGauge("eventTime-watermark",
+    s => convertStringDateToMillis(s.eventTime.get("watermark")), 0L)
```

Comment:
1. nit: rename `s` => `progress` to make it clear.
2. The eventTime-watermark metrics needs to be reported only if the map is not empty (event time). Could be skipped if the map is empty (processing time) to avoid confusion.

Reply:
1. will address
2. We don't know whether the map will be empty when calling `registerGauge`, and once we register the metric, `getValue` in Gauge is called from Dropwizard so I'm not sure we can control whether reporting the value or not.